File: hcs_toolbox_py/hpic_design.py
"""
Direct HPIC design (Homogeneous Proportional-Integral Control).

Original MATLAB: hpic_design.m

    u = uh + v,   dv/dt = ui

    uh = K0*x + nx^(1+mu) * K * d(-ln(nx)) * x
    ui = nx^(1+2*mu) * Ki * d(-ln(nx)) * x / (x' * d'(-ln(nx)) * P * Gd * d(-ln(nx)) * x)
"""

from .hpc_design import hpc_design
import numpy as np


def hpic_design(A, B, mu=-1.0, rho=1.0, gamma=0.0):
    """
    Design HPIC parameters directly.

    Parameters
    ----------
    A : (n, n) ndarray
        System matrix.
    B : (n, m) ndarray
        Control matrix.
    mu : float, optional
        Homogeneity degree (default -1).
    rho : float, optional
        Convergence time tuning parameter (default 1).
    gamma : float, optional
        Matched perturbation bound (default 0).

    Returns
    -------
    K0 : (m, n) ndarray
        Linear component gain.
    K : (m, n) ndarray
        Proportional nonlinear gain.
    Ki : (m, n) ndarray
        Integral gain.
    Gd : (n, n) ndarray
        Generator of dilation.
    P : (n, n) ndarray
        Positive definite matrix for homogeneous norm.
    """
    K0, K, Gd, P = hpc_design(A, B, mu, rho, gamma)
    Ki = -B.T @ P
    return K0, K, Ki, Gd, P

File: HCS_Toolbox_ver02/hpic_design.m
function [K0, K, Ki, Gd, P]=hpic_design(A, B, mu, rho, gamma)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% The function [K0, K, Ki, Gd, P]=hpic_design(A, B, mu) computes                    
%% parameters of Homogeneous Proportional Integral Controller (HPIC)     
%%                                                                      
%%     u=uh+v,  dv/dt=ui                          <- u= HPC + integral term  
%%
%%     uh=K0*x+nx^(1+mu)*K*d(-log(nx))*x          <- HPC 
%%
%%             nx^(1+2*mu)*Ki*d(-log(nx))*x
%%     ui=  ----------------------------------    <- integrant 
%%          x'*d'(-log(nx))*P*Gd*d(-log(nx))*x
%%                                                                      
%% The input parameters:                                             
%%     A  - system matrix (n x n)                                       
%%     B  - control matrix (n x m)               
%%     mu - homogeneity degree (mu=-1 corresponds to sliding mode control)       
%%                                                                      
%% The output parameters:                                                
%%     K0 - feedback gain (m x n) of the linear term of HPIC          
%%     K  - feedback gain (m x n) of the propotional term of HPIC  
%%     Ki - feedback gain (m x n) of the integral term of HPIC 
%%     Gd - anti-Hurwitz matrix (n x n) being a generator of the dilation 
%%                          d(s)=expm(s*Gd),   s - a real number                   
%%     P  - positive definite matrix which defines the weighted Euclidean 
%%         norm sqrt(x'*P*x) that induces the canonical homogeneous norm nx  
%%                                                                      
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
if nargin==4 [K0, K, Gd, P]=hpc_design(A,B,mu,rho);
elseif nargin==5 [K0, K, Gd, P]=hpc_design(A,B,mu,rho,gamma);
else [K0, K, Gd, P]=hpc_design(A,B,mu);
end;
Ki=-B'*P;

